File: shared/domain/tanda-strip.ts
// Cómo se LEE la tanda en vivo: 1 / 2 / 3, no solo la serie en curso.
// La ventana es la misma del riel (cerrada de antes, actual, siguiente).

export const TANDA_TODAS_HASTA = 4;
export const TANDA_VENTANA = 3;

export type TandaEstado = 'hecha' | 'actual' | 'futura' | 'saltada';

export type TandaPaso = {
  n: number;
  estado: TandaEstado;
};

export type TandaStrip = {
  pasos: readonly TandaPaso[];
  total: number;
};

export function tandaIndices(total: number, actual: number): number[] {
  if (total <= 0) return [];
  const cursor = Math.min(Math.max(0, actual), Math.max(0, total - 1));
  if (total <= TANDA_TODAS_HASTA) {
    return Array.from({ length: total }, (_, i) => i);
  }
  const ancho = Math.min(TANDA_VENTANA, total);
  const inicio = Math.min(Math.max(0, cursor - 1), Math.max(0, total - ancho));
  return Array.from({ length: ancho }, (_, k) => inicio + k);
}

export function tandaStrip(input: {
  total: number;
  actual: number;
  hechas: readonly number[];
  saltadas?: readonly number[];
}): TandaStrip {
  const hechas = new Set(input.hechas);
  const saltadas = new Set(input.saltadas ?? []);
  return {
    total: input.total,
    pasos: tandaIndices(input.total, input.actual).map((i) => {
      const estado: TandaEstado = saltadas.has(i)
        ? 'saltada'
        : hechas.has(i)
          ? 'hecha'
          : i === input.actual
            ? 'actual'
            : 'futura';
      return { n: i + 1, estado };
    }),
  };
}

export function tandaSeLee(strip: TandaStrip): string {
  return strip.pasos.map((p) => String(p.n)).join(' / ');
}

export function tandaEsVentana(strip: TandaStrip): boolean {
  return strip.pasos.length < strip.total;
}

File: ios/FAHYBRIKCore/Workout/TandaStrip.swift
import Foundation

/// Cómo se LEE la tanda en vivo. Espejo de `shared/domain/tanda-strip.ts`.
enum TandaStrip {
    static let todasHasta = 4
    static let ventana = 3

    enum Estado: String, Equatable {
        case hecha
        case actual
        case futura
        case saltada
    }

    struct Paso: Equatable {
        let n: Int
        let estado: Estado
    }

    struct Strip: Equatable {
        let pasos: [Paso]
        let total: Int

        var seLee: String { pasos.map { String($0.n) }.joined(separator: " / ") }
        var esVentana: Bool { pasos.count < total }
    }

    static func indices(total: Int, actual: Int) -> [Int] {
        guard total > 0 else { return [] }
        let cursor = min(max(0, actual), max(0, total - 1))
        if total <= todasHasta { return Array(0..<total) }
        let ancho = min(ventana, total)
        let inicio = min(max(0, cursor - 1), max(0, total - ancho))
        return Array(inicio..<(inicio + ancho))
    }

    static func strip(total: Int, actual: Int, hechas: Set<Int>, saltadas: Set<Int> = []) -> Strip {
        let pasos = indices(total: total, actual: actual).map { i -> Paso in
            let estado: Estado
            if saltadas.contains(i) { estado = .saltada }
            else if hechas.contains(i) { estado = .hecha }
            else if i == actual { estado = .actual }
            else { estado = .futura }
            return Paso(n: i + 1, estado: estado)
        }
        return Strip(pasos: pasos, total: total)
    }

    static func seLee(from series: [SetRecord], actual: Int) -> String {
        strip(from: series, actual: actual).seLee
    }

    static func strip(from series: [SetRecord], actual: Int) -> Strip {
        var hechas = Set<Int>()
        var saltadas = Set<Int>()
        for (i, rec) in series.enumerated() {
            if rec.status == "skipped" { saltadas.insert(i) }
            else if rec.confirmed { hechas.insert(i) }
        }
        return strip(total: series.count, actual: actual, hechas: hechas, saltadas: saltadas)
    }
}
